### Librerías

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt

pd.set_option('display.max_columns', 500)
pd.set_option('display.max_rows', 5000)

### Cargamos los datos

In [2]:
path_folder = './data/'
pd_fraude = pd.read_csv(path_folder + 'base.csv')

In [3]:
pd_fraude

fraud_bool  income  name_email_similarity  prev_address_months_count  \
0                1     0.9               0.166828                         -1   
1                1     0.9               0.296286                         -1   
2                1     0.9               0.044985                         -1   
3                1     0.9               0.159511                         -1   
4                1     0.9               0.596414                         -1   
...            ...     ...                    ...                        ...   
999995           0     0.6               0.192631                         -1   
999996           0     0.8               0.322989                        148   
999997           0     0.8               0.879403                         -1   
999998           0     0.9               0.762112                         -1   
999999           0     0.2               0.697452                         -1   

        current_address_months_count  customer_age  days_since_request  \
0                                 88            50            0.020925   
1                                144            50            0.005418   
2                                132            40            3.108549   
3                                 22            50            0.019079   
4                                218            50            0.004441   
...                              ...           ...                 ...   
999995                           104            40            0.030592   
999996                             9            50            1.628119   
999997                            30            20            0.018563   
999998                           189            20            0.015352   
999999                           321            20            2.655916   

        intended_balcon_amount payment_type  zip_count_4w   velocity_6h  \
0                    -1.331345           AA           769  10650.765523   
1                    -0.816224           AB           366    534.047319   
2                    -0.755728           AC           870   4048.534263   
3                    -1.205124           AB           810   3457.064063   
4                    -0.773276           AB           890   5020.341679   
...                        ...          ...           ...           ...   
999995               -1.044454           AB           804   7905.711839   
999996               -1.409803           AC          3306   5391.470463   
999997               34.692760           AA          1522   8063.102636   
999998               94.661055           AA          1418   8092.641762   
999999                9.908499           AA           951   6169.630036   

        velocity_24h  velocity_4w  bank_branch_count_8w  \
0        3134.319630  3863.647740                     1   
1        2670.918292  3124.298166                   718   
2        2893.621498  3159.590679                     1   
3        4054.908412  3022.261812                  1921   
4        2728.237159  3087.670952                  1990   
...              ...          ...                   ...   
999995   8341.468557  4972.635997                     1   
999996   4955.170808  5022.728108                     0   
999997   5670.654316  4377.196321                  2023   
999998   3982.582204  4394.803296                  1678   
999999   3695.308261  4352.334543                     2   

        date_of_birth_distinct_emails_4w employment_status  credit_risk_score  \
0                                      6                CA                185   
1                                      3                CA                259   
2                                     14                CB                177   
3                                      6                CA                110   
4                                      2                CA                295   
...                                  ...               ...    

### Análisis general de la tabla

Dimensión

In [4]:
print(pd_fraude.shape, pd_fraude.drop_duplicates().shape)

(1000000, 32) (1000000, 32)


El conjunto de datos no presenta información duplicada

Tipos de datos

In [5]:
pd_fraude.dtypes.to_dict()

{'fraud_bool': dtype('int64'),
 'income': dtype('float64'),
 'name_email_similarity': dtype('float64'),
 'prev_address_months_count': dtype('int64'),
 'current_address_months_count': dtype('int64'),
 'customer_age': dtype('int64'),
 'days_since_request': dtype('float64'),
 'intended_balcon_amount': dtype('float64'),
 'payment_type': dtype('O'),
 'zip_count_4w': dtype('int64'),
 'velocity_6h': dtype('float64'),
 'velocity_24h': dtype('float64'),
 'velocity_4w': dtype('float64'),
 'bank_branch_count_8w': dtype('int64'),
 'date_of_birth_distinct_emails_4w': dtype('int64'),
 'employment_status': dtype('O'),
 'credit_risk_score': dtype('int64'),
 'email_is_free': dtype('int64'),
 'housing_status': dtype('O'),
 'phone_home_valid': dtype('int64'),
 'phone_mobile_valid': dtype('int64'),
 'bank_months_count': dtype('int64'),
 'has_other_cards': dtype('int64'),
 'proposed_credit_limit': dtype('float64'),
 'foreign_request': dtype('int64'),
 'source': dtype('O'),
 'session_length_in_minutes': dty

### Exploración de la variable objetivo y tratamiento

In [6]:
pd_fraude_bool = pd_fraude['fraud_bool'].value_counts(normalize = True).mul(100).reset_index()
pd_fraude_bool['conteo'] = pd_fraude['fraud_bool'].value_counts()

pd_fraude_bool

fraud_bool  proportion  conteo
0           0     98.8971  988971
1           1      1.1029   11029

La variable objetivo está representada por 0 y 1. El valor 0 representa que no se ha cometido fraude, y la variable 1 que sí se ha cometido. En la muestra de datos, hay un 98.9% de instancias donde no se ha cometido fraude y un 1.1% donde se ha cometido.

### Estudio de los valores missing

Variables donde los valores missing se representan con un -1:
- prev_address_months_count
- current_address_months_count
- bank_months_count
- session_length_in_minutes
- device_distinct_emails

Variables donde los valores missing se representan con un número negativo:
- intended_balcon_amount

In [7]:
lista_nulos_dif = ['prev_address_months_count', 'current_address_months_count', 'bank_months_count',\
                  'session_length_in_minutes', 'device_distinct_emails_8w', 'intended_balcon_amount']

pd_series_null_columns = pd_fraude\
.drop(lista_nulos_dif, axis=1)\
.isnull().sum().sort_values(ascending=False)
pd_series_null_columns

fraud_bool                          0
income                              0
device_fraud_count                  0
keep_alive_session                  0
device_os                           0
source                              0
foreign_request                     0
proposed_credit_limit               0
has_other_cards                     0
phone_mobile_valid                  0
phone_home_valid                    0
housing_status                      0
email_is_free                       0
credit_risk_score                   0
employment_status                   0
date_of_birth_distinct_emails_4w    0
bank_branch_count_8w                0
velocity_4w                         0
velocity_24h                        0
velocity_6h                         0
zip_count_4w                        0
payment_type                        0
days_since_request                  0
customer_age                        0
name_email_similarity               0
month                               0
dtype: int64

Del resto de variables, vamos que no existen valores nulos entre sus datos

Vemos el número de valores missing para las demás variables

In [8]:
lista_nulos_neg_1 = ['prev_address_months_count', 'current_address_months_count', 'bank_months_count',\
                  'session_length_in_minutes', 'device_distinct_emails_8w']

pd_series_null_columns2 = (pd_fraude[lista_nulos_neg_1] == -1)\
.sum().sort_values(ascending=False)
pd_series_null_columns2

prev_address_months_count       712920
bank_months_count               253635
current_address_months_count      4254
session_length_in_minutes         2015
device_distinct_emails_8w          359
dtype: int64

In [9]:
pd_series_null_columns3 = (pd_fraude['intended_balcon_amount'] < 0).sum()
pd_series_null_columns3 = pd.Series(pd_series_null_columns3, index = ['intended_balcon_amount'])
pd_series_null_columns3

intended_balcon_amount    742523
dtype: int64

In [10]:
pd_null_columnas_final = pd.concat((pd_series_null_columns, pd_series_null_columns2, pd_series_null_columns3))\
.sort_values(ascending = False)
pd_null_columnas_final

intended_balcon_amount              742523
prev_address_months_count           712920
bank_months_count                   253635
current_address_months_count          4254
session_length_in_minutes             2015
device_distinct_emails_8w              359
income                                   0
month                                    0
name_email_similarity                    0
customer_age                             0
days_since_request                       0
payment_type                             0
zip_count_4w                             0
velocity_6h                              0
velocity_24h                             0
velocity_4w                              0
fraud_bool                               0
date_of_birth_distinct_emails_4w         0
employment_status                        0
credit_risk_score                        0
email_is_free                            0
housing_status                           0
phone_home_valid                         0
phone_mobil

In [11]:
pd_null_columnas = pd.DataFrame(pd_null_columnas_final, columns=['nulos_columnas']) 
pd_null_columnas['porcentaje_columnas'] = pd_null_columnas['nulos_columnas']/pd_fraude.shape[0]
pd_null_columnas

nulos_columnas  porcentaje_columnas
intended_balcon_amount                    742523             0.742523
prev_address_months_count                 712920             0.712920
bank_months_count                         253635             0.253635
current_address_months_count                4254             0.004254
session_length_in_minutes                   2015             0.002015
device_distinct_emails_8w                    359             0.000359
income                                         0             0.000000
month                                          0             0.000000
name_email_similarity                          0             0.000000
customer_age                                   0             0.000000
days_since_request                             0             0.000000
payment_type                                   0             0.000000
zip_count_4w                                   0             0.000000
velocity_6h                                    0             0.000000
velocity_24h                                   0             0.000000
velocity_4w                                    0             0.000000
fraud_bool                                     0             0.000000
date_of_birth_distinct_emails_4w               0             0.000000
employment_status                              0             0.000000
credit_risk_score                              0             0.000000
email_is_free                                  0             0.000000
housing_status                                 0             0.000000
phone_home_valid                               0             0.000000
phone_mobile_valid                             0             0.000000
has_other_cards                                0             0.000000
proposed_credit_limit                          0             0.000000
foreign_request                                0             0.000000
source                                         0             0.000000
device_os                                      0             0.000000
keep_alive_session                             0             0.000000
device_fraud_count                             0             0.000000
bank_branch_count_8w                           0             0.000000

Vemos que las variables 'intended_balcon_amount', 'prev_address_months_count' y 'bank_months_count' poseen una gran cantidad de valores missing, con un 74, 71 y 25% respectivamente de toda la muestra

### Tipos: Variables categóricas y numéricas

Agrupamos en una lista todas aquellas variables que son categóricas, ya sea porque lo indica directamente la información del datasheet, o por la propia naturaleza de la variable que siendo númerica se trata como categórica.

In [12]:
lista_variables_categoricas = [
    'payment_type',
    'employment_status',
    'email_is_free',
    'housing_status',
    'phone_home_valid',
    'phone_mobile_valid',
    'has_other_cards',
    'foreign_request',
    'source',
    'device_os',
    'keep_alive_session',
    'fraud_bool',
    'month'
]

Agrupamos el resto de variables que no son categóricas en una lista, pero primero agrupamos todas las varibales del dataset:

In [13]:
todas_las_variables = pd_fraude.columns
todas_las_variables

Index(['fraud_bool', 'income', 'name_email_similarity',
       'prev_address_months_count', 'current_address_months_count',
       'customer_age', 'days_since_request', 'intended_balcon_amount',
       'payment_type', 'zip_count_4w', 'velocity_6h', 'velocity_24h',
       'velocity_4w', 'bank_branch_count_8w',
       'date_of_birth_distinct_emails_4w', 'employment_status',
       'credit_risk_score', 'email_is_free', 'housing_status',
       'phone_home_valid', 'phone_mobile_valid', 'bank_months_count',
       'has_other_cards', 'proposed_credit_limit', 'foreign_request', 'source',
       'session_length_in_minutes', 'device_os', 'keep_alive_session',
       'device_distinct_emails_8w', 'device_fraud_count', 'month'],
      dtype='object')

In [14]:
lista_variables_numericas = []
for variable in todas_las_variables:
    if variable not in lista_variables_categoricas:
        lista_variables_numericas.append(variable)
lista_variables_numericas

['income',
 'name_email_similarity',
 'prev_address_months_count',
 'current_address_months_count',
 'customer_age',
 'days_since_request',
 'intended_balcon_amount',
 'zip_count_4w',
 'velocity_6h',
 'velocity_24h',
 'velocity_4w',
 'bank_branch_count_8w',
 'date_of_birth_distinct_emails_4w',
 'credit_risk_score',
 'bank_months_count',
 'proposed_credit_limit',
 'session_length_in_minutes',
 'device_distinct_emails_8w',
 'device_fraud_count']

Vemos qué tipo de variable numérica es cada una, si entera o continua

In [15]:
pd_fraude[lista_variables_numericas].dtypes.to_dict()

{'income': dtype('float64'),
 'name_email_similarity': dtype('float64'),
 'prev_address_months_count': dtype('int64'),
 'current_address_months_count': dtype('int64'),
 'customer_age': dtype('int64'),
 'days_since_request': dtype('float64'),
 'intended_balcon_amount': dtype('float64'),
 'zip_count_4w': dtype('int64'),
 'velocity_6h': dtype('float64'),
 'velocity_24h': dtype('float64'),
 'velocity_4w': dtype('float64'),
 'bank_branch_count_8w': dtype('int64'),
 'date_of_birth_distinct_emails_4w': dtype('int64'),
 'credit_risk_score': dtype('int64'),
 'bank_months_count': dtype('int64'),
 'proposed_credit_limit': dtype('float64'),
 'session_length_in_minutes': dtype('float64'),
 'device_distinct_emails_8w': dtype('int64'),
 'device_fraud_count': dtype('int64')}

### Preprocesamiento inicial de algunas variables

Vamos a guardar las variables que son categóricas como 'category'

In [16]:
for variable in lista_variables_categoricas:
    pd_fraude[variable] = pd_fraude[variable].astype('category')

In [17]:
pd_fraude[lista_variables_categoricas].dtypes

payment_type          category
employment_status     category
email_is_free         category
housing_status        category
phone_home_valid      category
phone_mobile_valid    category
has_other_cards       category
foreign_request       category
source                category
device_os             category
keep_alive_session    category
fraud_bool            category
month                 category
dtype: object

Tras observar las distintas variables del dataset, no he encontrado ninguna que puede ser útil para convertirla en el índice del dataframe.

Procedo a guardar en un csv los datos preprocesados

In [18]:
pd_fraude.to_csv('./data_preprocessing/pd_data_initial_preprocessing.csv')